business/dashboards.go

```diff
@@ -0,0 +1,148 @@
+package business
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+	"github.com/kiali/kiali/prometheus"
+)
+
+// DashboardsService deals with fetching dashboards from k8s client
+type DashboardsService struct {
+	prom prometheus.ClientInterface
+	mon  kubernetes.KialiMonitoringInterface
+}
+
+// NewDashboardsService initializes this business service
+func NewDashboardsService(mon kubernetes.KialiMonitoringInterface, prom prometheus.ClientInterface) DashboardsService {
+	return DashboardsService{prom: prom, mon: mon}
+}
+
+// GetDashboard returns a dashboard filled-in with target data
+func (in *DashboardsService) GetDashboard(params prometheus.CustomMetricsQuery, template string) (*models.MonitoringDashboard, error) {
+	dashboard, err := in.mon.GetDashboard(params.Namespace, template)
+	if err != nil {
```

Comment:
I don't get this logic (as external reviewer).
I understand that if first fetch fails 404/500 it goes to the default fetch in Kiali namespace.
Probably I'm missing something, but as this logic is pretty new I would suggest to add more comments just to follow the main flow of the logic.

Reply:
For example, I don't understand if query is asking for ns-1 and that doesn't exist, why it shouldn't return a 404 instead to fall into the istio-ns one.

Reply:
You're right that it should be commented.
This is an override mechanism. The reason is that we're going to provide some default dashboards that may not have to be installed in every namespaces. But those dashboards can be overriden per namespace, or user can create new dashboards in their app namespace.